src/mpi/misc/utils.c

```diff
@@ -197,3 +216,24 @@ int MPIR_Ilocalcopy(const void *sendbuf, MPI_Aint sendcount, MPI_Datatype sendty
   fn_fail:
     goto fn_exit;
 }
+
+/* Only single chunk copy (either sendtype or recvtype is contig) works with do_localcopy.
+ * For noncontig to noncontig, use MPIR_Localcopy_noncontig_stream.
```

Comment:
Did `MPIR_Localcopy_noncontig_stream` get removed from this PR? I'm not seeing it.

Reply:
Yeah, it was removed, and I forgot to update the comment. The noncontig localcopy will require another host bounce buffer which will involve a host to host copy. `MPIR_Typerep_(un)pack_stream` with host-to-host will just do synchronous copy, which will be incorrect since the data isn't ready yet. For non-contig case, I have to split the two copy between `MPIR_Typerep_(un)pack_stream` (for the gpu-host copy) and the host callback (for the host-host copy).

Reply:
I'm not sure I totally understand the need for host-host copy in your explanation. But looking at the code, is it safe to say that non-contig to non-contig stream-based localcopy is not supported in this PR?

Reply:
Refer to these lines -- https://github.com/pmodels/mpich/blob/107de6682de85383dd8c44e1980883b0f62f9566/src/mpi/stream/stream_enqueue.c#L550-L604

For contig datatypes, the macro just calls `MPIR_Localcopy_stream`. For non-contig datatypes, the macros split into two parts with a second part called from callbacks.

Reply:
> I'm not sure I totally understand the need for host-host copy in your explanation.

For noncontig localcopy from GPU->CPU, it is 
    GPU  [pack to] CPU [unpack to] CPU
, where the second step is CPU to CPU copy. Similarly for the localcopy from CPU -> GPU, there will be a host-to-host copy involved.

Reply:
Comments amended -- https://github.com/pmodels/mpich/blob/c9000a3802cfed7162cd078a938f73ef69fa7742/src/mpi/misc/utils.c#L220-L224

Reply:
👍 I mainly asked because I found the original `MPIR_Localcopy_noncontig_stream` API kind of clunky and unintuitive when compared to existing `MPIR_Localcopy` semantics. I'm happy there's a reasonable solution without it.